Add `match_only_text`, a space-efficient variant of `text`.

jpountz opened the pull request:
This adds a new `match_only_text` field, which indexes the same data as a `text`
field that has `index_options: docs` and `norms: false` and uses the `_source`
for positional queries like `match_phrase`.

Unlike `text`, this field doesn't support scoring and span queries.

This new field is part of the `text` family, so it is returned as a `text` field in the
`_field_caps` output.

Closes #64467

jpountz committed:
Add `match_only_text`, a space-efficient variant of `text`.

This adds a new `match_only_text` field, which indexes the same data as a `text`
field that has `index_options: docs` and `norms: false` and uses the `_source`
for positional queries like `match_phrase`. Unlike `text`, this field doesn't
support scoring.

An alternative to this new field could have been to make the `text` field still
able to run positional queries when positions are not indexed, but I like this
new field better because it avoids questions around how scoring should perform.

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

romseygeek commented on x-pack/plugin/mapper-match-only-text/src/main/java/org/elasticsearch/xpack/matchonlytext/mapper/MatchOnlyTextFieldMapper.java:
```diff
@@ -0,0 +1,302 @@
+    }
+
+    @Override
+    protected void doXContentBody(XContentBuilder builder, boolean includeDefaults, Params params) throws IOException {
```
I don't think you need to copy this bit, because there will be no BWC issue with a new mapper

romseygeek commented on x-pack/plugin/mapper-match-only-text/src/main/java/org/elasticsearch/xpack/matchonlytext/query/SourceConfirmedTextQuery.java:
```diff
@@ -0,0 +1,345 @@
+        }
+
+        private float computeFreq() throws IOException {
+            MemoryIndex index = new MemoryIndex();
```
Sneaky, I like it.  MemoryIndex always runs with ScoreMode.COMPLETE so it might be possible to speed things up a bit by calling `createIndexSearcher` and running against that with ScoreMode.COMPLETE_NO_SCORES.  Another trick that highlighters use when re-analyzing source is to add a FilteringTokenFilter to the MemoryIndex's analyzer that drops all tokens not included in the query - you need to be careful to catch Automata-based queries but QueryVisitor makes it not too difficult to implement.

jpountz replied:
This feels like a great idea for a follow-up PR? 0:)

romseygeek commented on x-pack/plugin/mapper-match-only-text/src/test/java/org/elasticsearch/xpack/matchonlytext/query/SourceConfirmedTextQueryTests.java:
```diff
@@ -0,0 +1,325 @@
+        }
+    }
+
+    public void testSpanNear() throws Exception {
```
Does this work, given the TODO in SourceConfirmedTextQuery implementation above?

jpountz replied:
Yes, the query just uses a MatchAllDocsQuery as an approximation, but the query "works".

jpountz replied:
Hopefully we'll create better approximations for spans in a follow-up.

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java:
```diff
@@ -741,8 +750,21 @@ public Query multiPhraseQuery(TokenStream stream, int slop, boolean enablePositi
         }
 
+        private int countTokens(TokenStream ts) throws IOException {
+            ts.reset();
```
maybe assert that this is a cached tokenstream?

jpountz replied:
this feels a bit restrictive, as it could be a CachedTokenStream wrapped in a filter?

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -670,7 +670,7 @@ private void merge(FieldMapper toMerge, Conflicts conflicts) {
         }
 
-        protected void toXContent(XContentBuilder builder, boolean includeDefaults) throws IOException {
+        public void toXContent(XContentBuilder builder, boolean includeDefaults) throws IOException {
```
I think you can avoid changing this, the concerns about BWC only apply to existing mappings and won't be a problem for entirely new field mappers.

romseygeek reviewed (commented):
This looks very cool! I have some suggestions around making the confirmation step a bit more efficient, although I guess there will be tradeoffs between the cost of building the index from source and how much work has to be done during analysis.

jpountz committed:
iter

romseygeek reviewed (approved):
LGTM, thanks @jpountz

jtibshirani commented on x-pack/plugin/mapper-match-only-text/src/main/java/org/elasticsearch/xpack/matchonlytext/mapper/MatchOnlyTextFieldMapper.java:
```diff
@@ -0,0 +1,289 @@
+                valueFetcher.setNextReader(context);
+                return docID -> {
+                    try {
+                        sourceLookup.setSegmentAndDocument(context, docID);
```
Should we use the `SourceLookup` from `QueryShardContext`? I guess the _source unlikely to be shared but we've been working on consolidating all document access through `QueryShardContext#searchLookup` (#66256).

jtibshirani commented on docs/reference/mapping/types.asciidoc:
```diff
@@ -69,6 +69,7 @@ values.
 ==== Text search types
 
 <<text,`text`>>:: Analyzed, unstructured text.
+<<match-only-text,`match_only_text`>>:: A more space-efficient variant of `text`.
```
We decided to make a single page for the `keyword` type family, so that users could easily compare the trade-offs between each type. between It'd be nice to do the same for the new 'text' family'.

jpountz committed:
Merge branch 'master' into feature/source_phrase_queries

jpountz committed:
Use source lookup from the shard context.

jpountz committed:
Update release version.

jpountz committed:
Consolidate docs with `text`.

jpountz commented:
Thanks for the helpful reviews @romseygeek and @jtibshirani.

jtibshirani reviewed (commented):
I left a couple last comments. One thing I wanted to double-check -- I guess some users could want scoring for most queries (as it can help surface relevant content) but confirm positional queries using _source? Maybe we're keeping things simple at first and assume this use case very often sorts on timestamp.

jpountz committed:
Fail phrase queries when _source is disabled.

jpountz committed:
Remove support for `store`.

jpountz committed:
Add tests for span and intervals queries.

jpountz committed:
Test for fuzzy query.

jpountz committed:
More tests.

jpountz commented:
> I guess some users could want scoring for most queries (as it can help surface relevant content) but confirm positional queries using _source?

I have long hesitated about this sort of things. I have considered the following options besides this PR:
 1. Keep norms and term frequencies indexed. If we did this we could have 100% scoring compatibility with the `text` field type by computing phrase frequencies from the _source instead of positions. I didn't like this option because I felt like most users wouldn't care enough about scoring to spend disk space on term frequencies and norms.
 2. Only index docs like the PR does but try to maintain some scoring. We can't have 100% scoring compatibility because most similarities need norms, and some of them also need total term frequencies, which are only computed when term frequencies are indexed. We could still have scoring enabled, e.g. by always using 1 for term frequencies and field lengths, but I worried that this might be more confusing than disabling scoring entirely since it would be different from the scoring that you get on `text` fields.

This field still provides some scoring in the sense that the score of a document will be equal with the number of matching terms from the query with this field. So if your query is a match query for `404 robots.txt`, documents that contain both terms will be returned before documents that contain only one of these terms when sorting by score. My assumption is that it would be good enough for the immense majority of users who have a Logging use-case.

jpountz commented:
I had initially thought I could make this field support span and interval queries, but this is more challenging than I thought because these query builders to not consult the field mappers to be constructed and just assume the field has been indexed with positions. We could fix this by adding new query builders in `MappedFieldType` but this would be a lot of work, and I believe we could go without span and intervals for the first version.

jtibshirani commented:
> This field still provides some scoring in the sense that the score of a document will be equal with the number of matching terms from the query with this field... My assumption is that it would be good enough for the immense majority of users who have a Logging use-case.

Thanks for the extra context, it does seem like this is the best approach -- it's simple to understand, saves the most space, and would satisfy most use cases.

> We could fix this by adding new query builders in MappedFieldType but this would be a lot of work, and I believe we could go without span and intervals for the first version.

This seems like a nice change for a follow-up. As I mentioned above, it'd be good to address so we don't have some queries that scan all documents.

jtibshirani reviewed (approved):
I only looked from a high level, but all my comments have been addressed. I like the name too!

jpountz committed:
Merge branch 'master' into feature/source_phrase_queries

jpountz committed:
Fix compilation.

jpountz committed:
iter

jpountz committed:
Merge branch 'master' into feature/source_phrase_queries

jpountz committed:
iter

jpountz committed:
iter

jpountz committed:
Merge branch 'master' into feature/source_phrase_queries

jpountz committed:
iter

jpountz committed:
iter

jpountz commented:
I finally took some time to come back to this PR. I moved the code from a new plugin in x-pack to `modules/mapper-extras` and changed license headers. I also added support for the `intervals` queries we have builders for in `MappedFieldType`, that is `match` and `prefix` queries, in order to further reduce the number of differences between `text` and `match_only_text`. @romseygeek Could you maybe review this part before I merge, specifically the `SourceIntervalsSource` class?

jpountz committed:
Fix compilation.

jtibshirani commented on docs/reference/mapping/types/match-only-text.asciidoc:
```diff
@@ -0,0 +1,73 @@
+
+[horizontal]
+
+<<analyzer,`analyzer`>>::
```
One thought I had since reviewing: if this is just targeted at log lines, would it make sense to cut down on analysis config options? For example, not allowing for a different search analyzer or search quote analyzers. Or even removing the option configuring the analyzer, just using a default that targets log lines. This could make it simpler to maintain long BWC for the field type. (This is a rough idea, and I am not sure it makes sense... maybe many users in fact tweak analyzers for log lines.)

jpountz replied:
Actually I think it's a good call, e.g. as far as I know, ECS doesn't configure analyzers. It would be much easier to add it in the future if it proves needed than to remove it when we want to ease backward compatibility.

jpountz committed:
Analysis is no longer configurable.

romseygeek commented on modules/mapper-extras/src/main/java/org/elasticsearch/index/query/SourceIntervalsSource.java:
```diff
@@ -0,0 +1,196 @@
+                return doc;
+            }
+
+            private boolean setIterator(int doc) {
```
I think this can throw IOException?

romseygeek commented on modules/mapper-extras/src/main/java/org/elasticsearch/index/query/SourceIntervalsSource.java:
```diff
@@ -0,0 +1,196 @@
+    public int hashCode() {
+        // Not using matchesProvider and valueFetcherProvider, which don't identify this source but are only used to avoid scanning linearly
+        // through all documents
+        return Objects.hash(in, indexAnalyzer);
```
The index analyzer should be immutable for an open index, I think? So I'm not sure that it needs to be included here or in equals.

jpountz replied:
Agreed. I'm including it because I want to avoid making too many assumptions about how this class is used by Elasticsearch. Is it fine with you?

romseygeek reviewed (commented):
The `IntervalsSource` parts look good to me.

jpountz committed:
iter

jpountz committed:
Merge branch 'master' into feature/source_phrase_queries

jpountz committed:
Intervals unit tests.

jpountz committed:
Fix docs now that `match_only_text` supports all interval queries.

jpountz committed:
Undo testing hack.

jpountz commented:
This field now fully supports intervals thanks to #71429.

jpountz committed:
Merge branch 'master' into feature/source_phrase_queries
